Adding Unit testing docs and reference page (#4603)
fishtown-analytics/docs.getdbt.com: website/docs/reference/resource-properties/unit-tests.md

---
title: "About unit tests property"
sidebar_label: "Unit tests"
resource_types: [models]
datatype: test
---

## Definition

Unit tests validate your SQL modeling logic on a small set of static inputs before you materialize your full model in production. They support a test-driven development approach, improving both the efficiency of developers and code reliability.

To run only your unit tests, use the command:
`dbt test --select test_type:unit`

<file name='dbt_project.yml'>

```yml

unit_tests:
  - name: <test-name> # this is the unique name of the test
    model: <model-name> 
      versions: #optional

> 	include: <list-of-versions-to-include> #optional

Reviewer: Not sure what this red block is

Author: I can't get rid of it no matter how I alter the spacing. I'm gonna write this off as a glitch until I can get into the preview to confirm

Reviewer 2: I used the GitHub UI to add spaces and replace it. I've seen it before but I'll have to Google how to prevent it from happening.

> 	include: <list-of-versions-to-include> #optional

Reviewer 2: ```suggestion
        include: <list-of-versions-to-include> #optional
```

> 	exclude: <list-of-versions-to-exclude> #optional

Reviewer 2: ```suggestion
        exclude: <list-of-versions-to-exclude> #optional
```

        rows:
          - {dictionary}
        fixture: <fixture-name>
      - input: ... # declare additional inputs
    expect:
      format: dict | csv
      # if format csv, either define dictionary of rows or name of fixture
      rows: 
        - {dictionary}
      fixture: <fixture-name>
    overrides: # optional: configuration for the dbt execution environment
      macros:
        is_incremental: true | false
        dbt_utils.current_timestamp: str
        # ... any other jinja function from https://docs.getdbt.com/reference/dbt-jinja-functions
        # ... any other context property
      vars: {dictionary}
      env_vars: {dictionary}
  - name: <test-name> ... # declare additional unit tests

  ```

</file>

### About writing unit tests

Unit tests are currently limited to testing SQL models and only models in your current project. When writing your unit tests, keep the following in mind:
- If your model has multiple versions, be default the unit test will run on *all* versions of your model. To specify version(s) of your model to unit test, `include` or `exclude` the desired versions in your model versions config:

 ```yaml

> # my test_is_valid_email_address unit test will run on all 3 versions

Reviewer: ```suggestion
# my test_is_valid_email_address unit test will run on all versions of my_model
```

unit-tests:
  - name: test_is_valid_email_address
    model: my_model
...
            

> # my test_is_valid_email_address unit test will run on ONLY version 2

Reviewer: ```suggestion
# my test_is_valid_email_address unit test will run on ONLY version 2 of my_model
```

unit-tests:
  - name: test_is_valid_email_address 
    model: my_model 
      versions:
        include: 
          - 2
...
            

> # my test_is_valid_email_address unit test will run on ONLY version 2 and 3

Reviewer: ```suggestion
# my test_is_valid_email_address unit test will run on all versions EXCEPT 1 of my_model
```

unit-tests:
  - name: test_is_valid_email_address
    model: my_model 
      versions:
        exclude: 
          - 1
...

```
            
-When using `format: dict` you must supply an in-line dictionary for `rows:` (this is the default, if you don’t specify a `format`)
    
```yml

unit_tests:
  - name: test_my_model
    model: my_model
    given:
      - input: ref('my_model_a')
        format: dict

>     		rows:

Reviewer: I think some of the formatting may have copied over weird from notion

          - {id: 1, name: gerda}
          - {id: 2, b: michelle}    

```

- When `format: csv`, can either supply:
  - An inline csv string for `rows:`
        
  ```yaml
  unit_tests:
    - name: test_my_model
      model: my_model
      given:
        - input: ref('my_model_a')
          format: csv
        	rows: |
            id,name
            1,gerda
            2,michelle

        ...
    ```
        
    - The name of a csv file in the `tests/fixtures` directory in your project (or the directory configured for [test-paths](https://docs.getdbt.com/reference/project-configs/test-paths)) for `fixture`:
        
    ```yaml
    unit_tests:
      - name: test_my_model
        model: my_model
        given:
          - input: ref('my_model_a')
            format: csv
            fixture: my_model_a_fixture
        ...
    ```
        

>     ```yaml

Reviewer: ```suggestion
    ```csv
```

    # tests/fixtures/my_model_a_fixture.csv
    1,gerda
    2,michelle
    ```
        
- `input:` string that represents a `ref` or `source` call:
    - `ref('my_model')` or `ref('my_model', v='2')` or `ref('dougs_project', 'users')`
    - `source('source_schema', 'source_name')`
- `input:` is optional for seeds:
    - If you don’t supply an input for a seed, we will use the seed *as* the input.
    - If you do supply an input for a seed, we will use that input instead.

> - You can also have “empty” inputs, by setting rows to an empty list `rows: []`:

Reviewer: ```suggestion
- You can also have “empty” inputs, by setting rows to an empty list `rows: []`
```

unit-tests:
  - name: test_is_valid_email_address
    model: my_model
...
            
# my test_is_valid_email_address unit test will run on ONLY version 2 of my_model
unit-tests:
  - name: test_is_valid_email_address 
    model: my_model 
      versions:
        include: 
          - 2
...
            
# my test_is_valid_email_address unit test will run on all versions EXCEPT 1 of my_model
unit-tests:
  - name: test_is_valid_email_address
    model: my_model 
      versions:
        exclude: 
          - 1
...

```
            
-When using `format: dict` you must supply an in-line dictionary for `rows:` (this is the default, if you don’t specify a `format`)
    
```yml

unit_tests:
  - name: test_my_model
    model: my_model
    given:
      - input: ref('my_model_a')
        format: dict
        rows:
          - {id: 1, name: gerda}
          - {id: 2, b: michelle}    

```

- When `format: csv`, can either supply:
  - An inline csv string for `rows:`
        
  ```yaml
  unit_tests:
    - name: test_my_model
      model: my_model
      given:
        - input: ref('my_model_a')
          format: csv

>         	rows: |

Author: ```suggestion
        rows: |
```

---
title: "About unit tests property"
sidebar_label: "Unit tests"
resource_types: [models]
datatype: test
---

> ## Definition

Reviewer 2: We can remove this. Our convention has been to use "Overview" but  I think we can use nothing in such a short page. We had discussed not having any H2s to clarify Google results.
```suggestion
```

> Unit tests validate your SQL modeling logic on a small set of static inputs before you materialize your full model in production. They support a test-driven development approach, improving both the efficiency of developers and code reliability.

Reviewer 2: Only a tiny sugg to make two benefits parallel:
```suggestion
Unit tests validate your SQL modeling logic on a small set of static inputs before you materialize your full model in production. They support a test-driven development approach, improving both the efficiency of developers and reliability of code.
```

    model: <model-name> 
      versions: #optional
        include: <list-of-versions-to-include> #optional
        exclude: <list-of-versions-to-exclude> #optional
    config: 
      meta: {dictionary}
      tags: <string> | [<string>]
    given:
      - input: <ref_or_source_call> # optional for seeds
        format: dict | csv
        # if format csv, either define dictionary of rows or name of fixture
        rows:
          - {dictionary}
        fixture: <fixture-name>
      - input: ... # declare additional inputs
    expect:
      format: dict | csv
      # if format csv, either define dictionary of rows or name of fixture
      rows: 
        - {dictionary}
      fixture: <fixture-name>
    overrides: # optional: configuration for the dbt execution environment
      macros:
        is_incremental: true | false
        dbt_utils.current_timestamp: str
        # ... any other jinja function from https://docs.getdbt.com/reference/dbt-jinja-functions
        # ... any other context property
      vars: {dictionary}
      env_vars: {dictionary}
  - name: <test-name> ... # declare additional unit tests

  ```

</file>

> ### About writing unit tests

Reviewer 2: You can move this up a level because the page title is H1.
```suggestion
## About writing unit tests
```

Unit tests are currently limited to testing SQL models and only models in your current project. When writing your unit tests, keep the following in mind:

> - If your model has multiple versions, be default the unit test will run on *all* versions of your model. To specify version(s) of your model to unit test, `include` or `exclude` the desired versions in your model versions config:

Reviewer 2: ```suggestion
- If your model has multiple versions, the default unit test will run on *all* versions of your model. To specify version(s) of your model to unit test, use `include` or `exclude` for the desired versions in your model versions config:
```

```yaml

# my test_is_valid_email_address unit test will run on all versions of my_model
unit-tests:
  - name: test_is_valid_email_address
    model: my_model
...
            
# my test_is_valid_email_address unit test will run on ONLY version 2 of my_model
unit-tests:
  - name: test_is_valid_email_address 
    model: my_model 
      versions:
        include: 
          - 2
...
            
# my test_is_valid_email_address unit test will run on all versions EXCEPT 1 of my_model
unit-tests:
  - name: test_is_valid_email_address
    model: my_model 
      versions:
        exclude: 
          - 1
...

```

>             

Reviewer 2: This might be worth making a section for clarity and easy perusing. Later, we can add more context under this title. 
```suggestion

### Format

```

> -When using `format: dict` you must supply an in-line dictionary for `rows:` (this is the default, if you don’t specify a `format`)

Reviewer 2: I'm a little confused about what `format: dict` does. Can we link to something or expand in a follow up PR?

Also this line doesn't need a bullet. 
```suggestion
When using `format: dict` you must supply an in-line dictionary for `rows:` (this is the default, if you don’t specify a `format`)
```

    
```yml

unit_tests:
  - name: test_my_model
    model: my_model
    given:
      - input: ref('my_model_a')
        format: dict
        rows:
          - {id: 1, name: gerda}
          - {id: 2, b: michelle}    

```

> - When `format: csv`, can either supply:

Reviewer 2: I think we can treat this like  a paragraph:
```suggestion
When `format: csv`, can either supply:
```

  - An inline csv string for `rows:`
        
  ```yaml
  unit_tests:
    - name: test_my_model
      model: my_model
      given:
        - input: ref('my_model_a')
          format: csv
          rows: |
            id,name
            1,gerda
            2,michelle

        ...
    ```

        
    - The name of a csv file in the `tests/fixtures` directory in your project (or the directory configured for [test-paths](https://docs.getdbt.com/reference/project-configs/test-paths)) for `fixture`:
        
    ```yaml
    unit_tests:
      - name: test_my_model
        model: my_model
        given:
          - input: ref('my_model_a')
            format: csv
            fixture: my_model_a_fixture
        ...
    ```
        
    ```csv
    # tests/fixtures/my_model_a_fixture.csv
    1,gerda
    2,michelle
    ```

>         

Reviewer 2: Same sugg here. 
```suggestion
        
### Input
        
```

- `input:` string that represents a `ref` or `source` call:
    - `ref('my_model')` or `ref('my_model', v='2')` or `ref('dougs_project', 'users')`
    - `source('source_schema', 'source_name')`
- `input:` is optional for seeds:
    - If you don’t supply an input for a seed, we will use the seed *as* the input.
    - If you do supply an input for a seed, we will use that input instead.
- You can also have “empty” inputs, by setting rows to an empty list `rows: []`

> Examples:

Reviewer 2: ```suggestion
### Examples
```

---

> title: "About unit tests property"

Reviewer: ![Screenshot 2024-02-14 at 11 01 24 AM](https://github.com/dbt-labs/docs.getdbt.com/assets/53586774/22c644a2-7f29-44b5-adc0-22b5b3423a91)

I can see this on version 1.7 - is that expected?

    model: <model-name> 
      versions: #optional
        include: <list-of-versions-to-include> #optional
        exclude: <list-of-versions-to-exclude> #optional
    config: 
      meta: {dictionary}
      tags: <string> | [<string>]
    given:
      - input: <ref_or_source_call> # optional for seeds
        format: dict | csv
        # if format csv, either define dictionary of rows or name of fixture
        rows:
          - {dictionary}
        fixture: <fixture-name>
      - input: ... # declare additional inputs
    expect:
      format: dict | csv
      # if format csv, either define dictionary of rows or name of fixture
      rows: 
        - {dictionary}
      fixture: <fixture-name>
    overrides: # optional: configuration for the dbt execution environment
      macros:
        is_incremental: true | false
        dbt_utils.current_timestamp: str
        # ... any other jinja function from https://docs.getdbt.com/reference/dbt-jinja-functions
        # ... any other context property
      vars: {dictionary}
      env_vars: {dictionary}
  - name: <test-name> ... # declare additional unit tests

  ```

</file>

## About writing unit tests

> Unit tests are currently limited to testing SQL models and only models in your current project. When writing your unit tests, keep the following in mind:

Reviewer: ```suggestion
Unit tests are currently limited to testing SQL models and only models in your current project. 
```

> - If your model has multiple versions, the default unit test will run on *all* versions of your model. To specify version(s) of your model to unit test, use `include` or `exclude` for the desired versions in your model versions config:

Reviewer: ```suggestion

### Versions
If your model has multiple versions, the default unit test will run on *all* versions of your model. To specify version(s) of your model to unit test, use `include` or `exclude` for the desired versions in your model versions config:
```

```yaml

# my test_is_valid_email_address unit test will run on all versions of my_model

> unit-tests:

Reviewer: ```suggestion
unit_tests:
```

  - name: test_is_valid_email_address
    model: my_model

> ...

Reviewer: ```suggestion
    ...
```

            
# my test_is_valid_email_address unit test will run on ONLY version 2 of my_model

> unit-tests:

Reviewer: ```suggestion
unit_tests:
```

  - name: test_is_valid_email_address 
    model: my_model 
      versions:
        include: 
          - 2

> ...

Reviewer: ```suggestion
    ...
```

            
# my test_is_valid_email_address unit test will run on all versions EXCEPT 1 of my_model

> unit-tests:

Reviewer: ```suggestion
unit_tests:
```

  - name: test_is_valid_email_address
    model: my_model 
      versions:
        exclude: 
          - 1

> ...

Reviewer: ```suggestion
    ...
```

```

### Format

When using `format: dict` you must supply an in-line dictionary for `rows:` (this is the default, if you don’t specify a `format`)
    
```yml

unit_tests:
  - name: test_my_model
    model: my_model
    given:
      - input: ref('my_model_a')
        format: dict
        rows:
          - {id: 1, name: gerda}
          - {id: 2, b: michelle}    

> 

Reviewer: ```suggestion
        ...
```

```

When `format: csv`, can either supply:
  - An inline csv string for `rows:`
        
  ```yaml
  unit_tests:
    - name: test_my_model
      model: my_model
      given:
        - input: ref('my_model_a')
          format: csv
          rows: |
            id,name
            1,gerda
            2,michelle

> 

Reviewer: ```suggestion
```

        ...
    ```

        

>     - The name of a csv file in the `tests/fixtures` directory in your project (or the directory configured for [test-paths](https://docs.getdbt.com/reference/project-configs/test-paths)) for `fixture`:

Reviewer: ![Screenshot 2024-02-14 at 10 50 35 AM](https://github.com/dbt-labs/docs.getdbt.com/assets/53586774/a1245a3c-bbde-4a1c-80d7-9d1cb9c5321c)

This bullet should be at the same indentation as "An inline csv string for `rows:`"

Reviewer: @matthewshaver do you know how to fix this?

Author: Yup! I'll line it up!

        
    ```yaml
    unit_tests:
      - name: test_my_model
        model: my_model
        given:
          - input: ref('my_model_a')
            format: csv
            fixture: my_model_a_fixture
        ...
    ```
        
    ```csv
    # tests/fixtures/my_model_a_fixture.csv
    1,gerda
    2,michelle
    ```
        
### Input
        
- `input:` string that represents a `ref` or `source` call:
    - `ref('my_model')` or `ref('my_model', v='2')` or `ref('dougs_project', 'users')`
    - `source('source_schema', 'source_name')`
- `input:` is optional for seeds:
    - If you don’t supply an input for a seed, we will use the seed *as* the input.
    - If you do supply an input for a seed, we will use that input instead.
- You can also have “empty” inputs, by setting rows to an empty list `rows: []`

> ### Examples

Reviewer: ```suggestion
## Examples
```

```yml

unit_tests:
  - name: test_is_valid_email_address # this is the unique name of the test

>     model: my_model # name of the model I'm unit testing

Reviewer: ```suggestion
    model: dim_customers # name of the model I'm unit testing
```

    given: # the mock data for your inputs

>       - input: ref('users')

Reviewer: ```suggestion
      - input: ref('stg_customers')
```

        rows:

>          - {user_id: 1, email: cool@example.com,     email_top_level_domain: example.com}

Reviewer: ```suggestion
         - {customer_id: 1, email: cool@example.com,     email_top_level_domain: example.com}
```

>          - {user_id: 2, email: cool@unknown.com,     email_top_level_domain: unknown.com}

Reviewer: ```suggestion
         - {customer_id: 2, email: cool@unknown.com,     email_top_level_domain: unknown.com}
```

>          - {user_id: 3, email: badgmail.com,         email_top_level_domain: gmail.com}

Reviewer: ```suggestion
         - {customer_id: 3, email: badgmail.com,         email_top_level_domain: gmail.com}
```

>          - {user_id: 4, email: missingdot@gmailcom,  email_top_level_domain: gmail.com}

Reviewer: ```suggestion
         - {customer_id: 4, email: missingdot@gmailcom,  email_top_level_domain: gmail.com}
```

>       - input: ref('top_level_domains')

Reviewer: ```suggestion
      - input: ref('top_level_email_domains')
```

        rows:
         - {tld: example.com}
         - {tld: gmail.com}
    expect: # the expected output given the inputs above
      rows:

>         - {user_id: 1, is_valid_email_address: true}

Reviewer: ```suggestion
        - {customer_id: 1, is_valid_email_address: true}
```

>         - {user_id: 2, is_valid_email_address: false}

Reviewer: ```suggestion
        - {customer_id: 2, is_valid_email_address: false}
```

>         - {user_id: 3, is_valid_email_address: false}

Reviewer: ```suggestion
        - {customer_id: 3, is_valid_email_address: false}
```

>         - {user_id: 4, is_valid_email_address: false}

Reviewer: ```suggestion
        - {customer_id: 4, is_valid_email_address: false}
```

```

```yml

unit_tests:
  - name: test_is_valid_email_address # this is the unique name of the test

>     model: my_model # name of the model I'm unit testing

Reviewer: ```suggestion
    model: dim_customers # name of the model I'm unit testing
```

    given: # the mock data for your inputs

>       - input: ref('users')

Reviewer: ```suggestion
      - input: ref('stg_customers')
```

        rows:

>          - {user_id: 1, email: cool@example.com,     email_top_level_domain: example.com}

Reviewer: ```suggestion
         - {customer_id: 1, email: cool@example.com,     email_top_level_domain: example.com}
```

>          - {user_id: 2, email: cool@unknown.com,     email_top_level_domain: unknown.com}

Reviewer: ```suggestion
         - {customer_id: 2, email: cool@unknown.com,     email_top_level_domain: unknown.com}
```

>          - {user_id: 3, email: badgmail.com,         email_top_level_domain: gmail.com}

Reviewer: ```suggestion
         - {customer_id: 3, email: badgmail.com,         email_top_level_domain: gmail.com}
```

>          - {user_id: 4, email: missingdot@gmailcom,  email_top_level_domain: gmail.com}

Reviewer: ```suggestion
         - {customer_id: 4, email: missingdot@gmailcom,  email_top_level_domain: gmail.com}
```

>       - input: ref('top_level_domains')

Reviewer: ```suggestion
      - input: ref('top_level_email_domains')
```

        format: csv
        rows: |
          tld
          example.com
          gmail.com
    expect: # the expected output given the inputs above
      format: csv

>       fixture: my_model_fixture

Reviewer: ```suggestion
      fixture: valid_email_address_fixture_output
```

---
title: "About unit tests property"
sidebar_label: "unit tests"
resource_types: [models]
datatype: test
---

> 

Reviewer: Hmm - do we think this is a property? It's more like a new resource type

Author: @graciegoheen I just cribbed from data tests which are listed as a property. Should I change just unit tests or both?

Reviewer: I'm not actually sure! Just looking at all other pages in this section, they look like things you apply to the model/columns in your yml config:
```
models:
  - name: <model_name>
    config:
```

vs. unit tests which is a top level yml 
```
unit_tests:
...
```

What do you think? How do you define a "property"?

Reviewer: Just checking back in about this! I think we decided to move this to a new "For unit tests" section (like how exposures is)

Reviewer 2: > @matthewshaver is this updated re: our convo a couple of weeks ago? I _think_ we had decided to make this more like the "For exposures" property section.
> 
...
> 
> We may want to go ahead and rename the "For tests" section to "For data tests" as well. Lemme know what you think!
> 
> cc: @runleonarun in case you have thoughts here!

Hey @graciegoheen Sorry I missed the previous @mention. That makes sense to me. Are Unit tests like exposures in that they are "special properties?"  or are they [straight-up configs](https://docs.getdbt.com/reference/configs-and-properties#where-can-i-define-configs) that we can add to the dbt_project.yml? The example suggests configs because it's part of a project.yml file and top-level, but just want to double check. 

>Exposure properties are "special properties" in that you can't configure them in the dbt_project.yml file or using config() blocks. Refer to [Configs and properties](https://docs.getdbt.com/reference/configs-and-properties#which-properties-are-not-also-configs) for more info.

Author: I also added the "For unit Tests" folder to the Reference page which contains just the unit test file.

I also added a "Tests" folder to the /docs site under "Build your DAG" which contains unit test and data test files

Reviewer: I like how this looks!

<file name='dbt_project.yml'>

```yml

> unit-tests:

Reviewer: ```suggestion
unit_tests:
```

We recently changed this to be _ everywhere!

  - name: <test-name> # this is the unique name of the test
    model: <model-name> 
    config: 

> 			meta: {dictionary}

Reviewer: I think some weird stuff happened with spacing here

sidebar_label: "unit tests"
resource_types: [models]
datatype: test
---

<file name='dbt_project.yml'>

```yml

unit_tests:
  - name: <test-name> # this is the unique name of the test
    model: <model-name> 
    config: 
      meta: {dictionary}
      tags: <string> | [<string>]
    given:
      - input: <ref_or_source_call> # optional for seeds
        format: dict | csv
        # if format csv, either define dictionary of rows or name of fixture
        rows:
          - {dictionary}
        fixture: <fixture-name>
      - input: ... # declare additional inputs
    expect:
      format: dict | csv
      # if format csv, either define dictionary of rows or name of fixture
      rows: 
        - {dictionary}
      fixture: <fixture-name>
    overrides: # optional: configuration for the dbt execution environment
      macros:
        is_incremental: true | false
        dbt_utils.current_timestamp: str
        # ... any other jinja function from https://docs.getdbt.com/reference/dbt-jinja-functions
        # ... any other context property
      vars: {dictionary}
      env_vars: {dictionary}
  - name: <test-name> ... # declare additional unit tests

  ```

</file>

> ## Definition

Reviewer 2: I don't think we need this title if we move the definition up.
```suggestion
```

Author: This is the format that we use for the other pages. Should we make a note to change all of them?

Reviewer 2: @matthewshaver Yes definitely. I've been doing it little by little but this would be a great thing to change in one fell swoop.

> Unit tests validate your modeling logic on a small set of static inputs before you materialize your full model in production. Unit tests enable test-driven development, with benefits for developer efficiency and code reliability. 

Reviewer 2: I suggest moving the definition to the top so people can see it in google searches

> Unit tests validate your modeling logic on a small set of static inputs before you materialize your full model in production. Unit tests enable test-driven development, with benefits for developer efficiency and code reliability. 

Reviewer 3: ```suggestion
Unit tests validate your SQL modeling logic on a small set of static inputs before you materialize your full model in production. They support a test-driven development approach, improving both the efficiency of developers and code reliability.
```

Author: @graciegoheen do we want to use `SQL` here or leave it as "modeling logic"?

Reviewer: I think SQL is a good addition, given we don't currently support unit tests on python models

> To run just your unit tests:

Reviewer 3: ```suggestion
To execute only your unit tests, use the command:
```

> To run just your unit tests:

Author: Split the difference:
```suggestion
To run only your unit tests, use the command:
```

> `dbt test —-select test_type:unit`

Reviewer 3: the dashes were looking weird in the preview:
![Screenshot 2024-01-10 at 12 17 37](https://github.com/dbt-labs/docs.getdbt.com/assets/89008547/2f096f58-2ca9-4083-b933-63ad939df22e)

```suggestion
`dbt test --select test_type:unit`
```
